File: src-tauri/macos-widget/Sources/WidgetSnapshotContract/WidgetSnapshot.swift
import Foundation

public let widgetSnapshotSchemaVersion = 1

public enum WidgetTimelinePhase: String, Codable, Sendable {
    case idle
    case before
    case working
    case `break`
    case done
}

public enum WidgetCountdownKind: String, Codable, Sendable {
    case none
    case shiftStarts
    case workRemaining
    case breakEnds
    case complete
}

public struct WidgetShiftSegment: Codable, Equatable, Sendable {
    public let startAtMs: Int64
    public let endAtMs: Int64

    public init(startAtMs: Int64, endAtMs: Int64) {
        self.startAtMs = startAtMs
        self.endAtMs = endAtMs
    }
}

public struct WidgetShiftTimeline: Codable, Equatable, Sendable {
    public let segments: [WidgetShiftSegment]
    public let plannedEndAtMs: Int64
    public let overtimeEndAtMs: Int64?

    public init(
        segments: [WidgetShiftSegment],
        plannedEndAtMs: Int64,
        overtimeEndAtMs: Int64?
    ) {
        self.segments = segments
        self.plannedEndAtMs = plannedEndAtMs
        self.overtimeEndAtMs = overtimeEndAtMs
    }
}

public struct WidgetTimelineEntry: Codable, Equatable, Sendable {
    public let dateMs: Int64
    public let validUntilMs: Int64
    public let phase: WidgetTimelinePhase
    public let labelKey: String
    public let countdownKind: WidgetCountdownKind
    public let countdownValueAtDateMs: Int64
    public let countdownTargetAtMs: Int64?
    public let remainingEffectiveMsAtDateMs: Int64
    public let progressAtDate: Double
    public let nextBoundaryAtMs: Int64?

    public init(
        dateMs: Int64,
        validUntilMs: Int64,
        phase: WidgetTimelinePhase,
        labelKey: String,
        countdownKind: WidgetCountdownKind,
        countdownValueAtDateMs: Int64,
        countdownTargetAtMs: Int64?,
        remainingEffectiveMsAtDateMs: Int64,
        progressAtDate: Double,
        nextBoundaryAtMs: Int64?
    ) {
        self.dateMs = dateMs
        self.validUntilMs = validUntilMs
        self.phase = phase
        self.labelKey = labelKey
        self.countdownKind = countdownKind
        self.countdownValueAtDateMs = countdownValueAtDateMs
        self.countdownTargetAtMs = countdownTargetAtMs
        self.remainingEffectiveMsAtDateMs = remainingEffectiveMsAtDateMs
        self.progressAtDate = progressAtDate
        self.nextBoundaryAtMs = nextBoundaryAtMs
    }

    /// Advances a precomputed interval to a later presentation time. Schedule
    /// boundaries still come from the producer; this only projects the linear
    /// values inside one already-resolved interval.
    public func projected(atMs nowMs: Int64) -> WidgetTimelineEntry {
        #if os(iOS)
        let projectsProgress = phase == .working || phase == .before
        #else
        // The macOS product keeps its existing pre-shift presentation. DoneAt
        // countdown progress is an iOS-only surface change.
        let projectsProgress = phase == .working
        #endif
        guard projectsProgress,
              nowMs > dateMs,
              progressAtDate < 100,
              remainingEffectiveMsAtDateMs > 0
        else {
            return self
        }

        let projectedDateMs = min(nowMs, validUntilMs)
        let elapsedMs = min(
            remainingEffectiveMsAtDateMs,
            max(0, projectedDateMs - dateMs)
        )
        let remainingMs = max(0, remainingEffectiveMsAtDateMs - elapsedMs)
        let remainingFraction = max(0.000_001, 1 - progressAtDate / 100)
        let totalEffectiveMs = Double(remainingEffectiveMsAtDateMs) / remainingFraction
        let progress = min(
            100,
            max(progressAtDate, progressAtDate + Double(elapsedMs) / totalEffectiveMs * 100)
        )

        return WidgetTimelineEntry(
            dateMs: projectedDateMs,
            validUntilMs: validUntilMs,
            phase: phase,
            labelKey: labelKey,
            countdownKind: countdownKind,
            countdownValueAtDateMs: max(0, countdownValueAtDateMs - elapsedMs),
            countdownTargetAtMs: countdownTargetAtMs,
            remainingEffectiveMsAtDateMs: remainingMs,
            progressAtDate: progress,
            nextBoundaryAtMs: nextBoundaryAtMs
        )
    }
}

/// One row of the in-app "coming up" list, projected salary-free for large
/// Home Screen widgets. Titles and details are already localized by the
/// producer; the extension only filters by time and picks a glyph.
public struct WidgetUpcomingItem: Codable, Equatable, Sendable, Identifiable {
    public let id: String
    public let kind: String
    public let title: String
    public let detail: String
    public let dateMs: Int64

    public init(id: String, kind: String, title: String, detail: String, dateMs: Int64) {
        self.id = id
        self.kind = kind
        self.title = title
        self.detail = detail
        self.dateMs = dateMs
    }
}

public struct WidgetSnapshot: Codable, Equatable, Sendable {
    public let schemaVersion: Int
    public let generatedAtMs: Int64
    public let expiresAtMs: Int64
    public let locale: String
    public let shift: WidgetShiftTimeline?
    public let entries: [WidgetTimelineEntry]
    /// Events still ahead of `generatedAtMs`, used by iOS large / extra-large
    /// widgets. Recurring snapshots persist the remaining shifts in the
    /// validity window so a 12-hour WidgetKit reread still has rows after
    /// the original 36-hour presentation window. Absent from older snapshots;
    /// the decoder treats that as empty so macOS and the TypeScript fixture
    /// stay on schema 1.
    public let upcoming: [WidgetUpcomingItem]

    public init(
        schemaVersion: Int,
        generatedAtMs: Int64,
        expiresAtMs: Int64,
        locale: String,
        shift: WidgetShiftTimeline?,
        entries: [WidgetTimelineEntry],
        upcoming: [WidgetUpcomingItem] = []
    ) {
        self.schemaVersion = schemaVersion
        self.generatedAtMs = generatedAtMs
        self.expiresAtMs = expiresAtMs
        self.locale = locale
        self.shift = shift
        self.entries = entries
        self.upcoming = upcoming
    }

    enum CodingKeys: String, CodingKey {
        case schemaVersion, generatedAtMs, expiresAtMs, locale, shift, entries, upcoming
    }

    public init(from decoder: Decoder) throws {
        let container = try decoder.container(keyedBy: CodingKeys.self)
        schemaVersion = try container.decode(Int.self, forKey: .schemaVersion)
        generatedAtMs = try container.decode(Int64.self, forKey: .generatedAtMs)
        expiresAtMs = try container.decode(Int64.self, forKey: .expiresAtMs)
        locale = try container.decode(String.self, forKey: .locale)
        shift = try container.decodeIfPresent(WidgetShiftTimeline.self, forKey: .shift)
        entries = try container.decode([WidgetTimelineEntry].self, forKey: .entries)
        upcoming = try container.decodeIfPresent([WidgetUpcomingItem].self, forKey: .upcoming) ?? []
    }

    public func encode(to encoder: Encoder) throws {
        var container = encoder.container(keyedBy: CodingKeys.self)
        try container.encode(schemaVersion, forKey: .schemaVersion)
        try container.encode(generatedAtMs, forKey: .generatedAtMs)
        try container.encode(expiresAtMs, forKey: .expiresAtMs)
        try container.encode(locale, forKey: .locale)
        try container.encodeIfPresent(shift, forKey: .shift)
        try container.encode(entries, forKey: .entries)
        if !upcoming.isEmpty {
            try container.encode(upcoming, forKey: .upcoming)
        }
    }

    /// Selects a frontend-precomputed interval. This intentionally contains no
    /// workday, break, overtime, or cross-midnight rules.
    public func entry(atMs nowMs: Int64) -> WidgetTimelineEntry? {
        guard schemaVersion == widgetSnapshotSchemaVersion,
              nowMs >= generatedAtMs,
              nowMs < expiresAtMs
        else {
            return nil
        }

        return entries.first {
            nowMs >= $0.dateMs && nowMs < $0.validUntilMs
        }
    }

    /// Expands only the linear presentation of working intervals. It never
    /// invents schedule boundaries: those remain the producer's responsibility.
    public func presentationEntries(
        startingAtMs nowMs: Int64,
        progressStepMs: Int64,
        endingAtMs requestedEndMs: Int64? = nil
    ) -> [WidgetTimelineEntry] {
        let presentationEndMs = min(expiresAtMs, requestedEndMs ?? expiresAtMs)
        guard schemaVersion == widgetSnapshotSchemaVersion,
              progressStepMs > 0,
              nowMs >= generatedAtMs,
              nowMs < presentationEndMs
        else {
            return []
        }

        var result: [WidgetTimelineEntry] = []
        for source in entries where source.validUntilMs > nowMs && source.dateMs < presentationEndMs {
            let firstDateMs = max(nowMs, source.dateMs)
            result.append(source.projected(atMs: firstDateMs))

            guard source.phase == .working else { continue }
            var updateAtMs = firstDateMs + progressStepMs
            let intervalEndMs = min(source.validUntilMs, presentationEndMs)
            while updateAtMs < intervalEndMs {
                result.append(source.projected(atMs: updateAtMs))
                updateAtMs += progressStepMs
            }
        }
        return result.sorted { $0.dateMs < $1.dateMs }
    }
}

File: src-mobile/android/core/domain/src/main/kotlin/com/rainif/doneat/core/domain/widget/WidgetSnapshot.kt
package com.rainif.doneat.core.domain.widget

import kotlinx.serialization.json.Json
import kotlinx.serialization.json.JsonArray
import kotlinx.serialization.json.JsonNull
import kotlinx.serialization.json.JsonObject
import kotlinx.serialization.json.JsonPrimitive
import kotlinx.serialization.json.buildJsonObject
import kotlinx.serialization.json.double
import kotlinx.serialization.json.int
import kotlinx.serialization.json.jsonArray
import kotlinx.serialization.json.jsonObject
import kotlinx.serialization.json.jsonPrimitive
import kotlinx.serialization.json.long
import kotlinx.serialization.json.longOrNull

/**
 * The widget's whole world (the shared `WidgetSnapshotContract`, schema 1):
 * precomputed intervals the widget only picks from by time. It holds no
 * schedule rules and nothing about money: no salary, earnings or records.
 */
const val WIDGET_SNAPSHOT_SCHEMA = 1

enum class WidgetPhase(val key: String) { IDLE("idle"), BEFORE("before"), WORKING("working"), BREAK("break"), DONE("done") }

enum class WidgetCountdownKind(val key: String) {
    NONE("none"), SHIFT_STARTS("shiftStarts"), WORK_REMAINING("workRemaining"), BREAK_ENDS("breakEnds"), COMPLETE("complete"),
}

data class WidgetEntry(
    val dateMs: Long,
    val validUntilMs: Long,
    val phase: WidgetPhase,
    /** A string resource name: "widgetWorking", "lunchInProgress", "offWorkToday"… */
    val labelKey: String,
    val countdownKind: WidgetCountdownKind,
    val countdownValueAtDateMs: Long,
    /** The wall-clock moment counted toward: work start, work resuming, clock-off. */
    val countdownTargetAtMs: Long?,
    val remainingEffectiveMsAtDateMs: Long,
    /** 0…100. */
    val progressAtDate: Double,
    val nextBoundaryAtMs: Long?,
) {
    /**
     * This interval at a later moment inside it (the contract's `projected`,
     * iOS behaviour): working and pre-shift intervals advance linearly; the
     * boundaries themselves always come from the producer.
     */
    fun projected(nowMs: Long): WidgetEntry {
        val advances = phase == WidgetPhase.WORKING || phase == WidgetPhase.BEFORE
        if (!advances || nowMs <= dateMs || progressAtDate >= 100 || remainingEffectiveMsAtDateMs <= 0) return this
        val at = minOf(nowMs, validUntilMs)
        val elapsed = minOf(remainingEffectiveMsAtDateMs, maxOf(0, at - dateMs))
        val remainingFraction = maxOf(0.000_001, 1 - progressAtDate / 100)
        val total = remainingEffectiveMsAtDateMs / remainingFraction
        return copy(
            dateMs = at,
            countdownValueAtDateMs = maxOf(0, countdownValueAtDateMs - elapsed),
            remainingEffectiveMsAtDateMs = maxOf(0, remainingEffectiveMsAtDateMs - elapsed),
            progressAtDate = minOf(100.0, maxOf(progressAtDate, progressAtDate + elapsed / total * 100)),
        )
    }

    /**
     * When the on-screen countdown reaches zero. Remaining effective work,
     * not the wall-clock finish: counting to 19:00 would eat a lunch gap.
     */
    val timerEndAtMs: Long?
        get() = when (countdownKind) {
            WidgetCountdownKind.WORK_REMAINING -> remainingEffectiveMsAtDateMs.takeIf { it > 0 }?.let { dateMs + it }
            else -> countdownTargetAtMs
        }
}

/** One "coming up" row, already worded by the app. */
data class WidgetUpcomingItem(val id: String, val kind: String, val title: String, val detail: String, val dateMs: Long)

data class WidgetSnapshot(
    val schemaVersion: Int,
    val generatedAtMs: Long,
    val expiresAtMs: Long,
    val locale: String,
    val entries: List<WidgetEntry>,
    val upcoming: List<WidgetUpcomingItem> = emptyList(),
) {
    /** The interval covering [nowMs], or null when the snapshot is stale or from another schema. */
    fun entry(nowMs: Long): WidgetEntry? {
        if (schemaVersion != WIDGET_SNAPSHOT_SCHEMA || nowMs < generatedAtMs || nowMs >= expiresAtMs) return null
        return entries.firstOrNull { nowMs >= it.dateMs && nowMs < it.validUntilMs }?.projected(nowMs)
    }

    fun encode(): String = buildJsonObject {
        put("schemaVersion", JsonPrimitive(schemaVersion))
        put("generatedAtMs", JsonPrimitive(generatedAtMs))
        put("expiresAtMs", JsonPrimitive(expiresAtMs))
        put("locale", JsonPrimitive(locale))
        put("entries", JsonArray(entries.map { e ->
            buildJsonObject {
                put("dateMs", JsonPrimitive(e.dateMs))
                put("validUntilMs", JsonPrimitive(e.validUntilMs))
                put("phase", JsonPrimitive(e.phase.key))
                put("labelKey", JsonPrimitive(e.labelKey))
                put("countdownKind", JsonPrimitive(e.countdownKind.key))
                put("countdownValueAtDateMs", JsonPrimitive(e.countdownValueAtDateMs))
                put("countdownTargetAtMs", e.countdownTargetAtMs?.let(::JsonPrimitive) ?: JsonNull)
                put("remainingEffectiveMsAtDateMs", JsonPrimitive(e.remainingEffectiveMsAtDateMs))
                put("progressAtDate", JsonPrimitive(e.progressAtDate))
                put("nextBoundaryAtMs", e.nextBoundaryAtMs?.let(::JsonPrimitive) ?: JsonNull)
            }
        }))
        put("upcoming", JsonArray(upcoming.map { u ->
            buildJsonObject {
                put("id", JsonPrimitive(u.id))
                put("kind", JsonPrimitive(u.kind))
                put("title", JsonPrimitive(u.title))
                put("detail", JsonPrimitive(u.detail))
                put("dateMs", JsonPrimitive(u.dateMs))
            }
        }))
    }.toString()

    companion object {
        /** Null for anything unreadable: the widget then shows its "open the app" state. */
        fun decode(text: String): WidgetSnapshot? = runCatching {
            val o = Json.parseToJsonElement(text).jsonObject
            fun JsonObject.long(key: String) = getValue(key).jsonPrimitive.long
            fun JsonObject.string(key: String) = getValue(key).jsonPrimitive.content
            fun JsonObject.optionalLong(key: String) = (get(key) as? JsonPrimitive)?.longOrNull
            WidgetSnapshot(
                schemaVersion = o.getValue("schemaVersion").jsonPrimitive.int,
                generatedAtMs = o.long("generatedAtMs"),
                expiresAtMs = o.long("expiresAtMs"),
                locale = o.string("locale"),
                entries = o.getValue("entries").jsonArray.map { element ->
                    val e = element.jsonObject
                    WidgetEntry(
                        dateMs = e.long("dateMs"),
                        validUntilMs = e.long("validUntilMs"),
                        phase = WidgetPhase.entries.first { it.key == e.string("phase") },
                        labelKey = e.string("labelKey"),
                        countdownKind = WidgetCountdownKind.entries.first { it.key == e.string("countdownKind") },
                        countdownValueAtDateMs = e.long("countdownValueAtDateMs"),
                        countdownTargetAtMs = e.optionalLong("countdownTargetAtMs"),
                        remainingEffectiveMsAtDateMs = e.long("remainingEffectiveMsAtDateMs"),
                        progressAtDate = e.getValue("progressAtDate").jsonPrimitive.double,
                        nextBoundaryAtMs = e.optionalLong("nextBoundaryAtMs"),
                    )
                },
                upcoming = (o["upcoming"] as? JsonArray).orEmpty().map { element ->
                    val u = element.jsonObject
                    WidgetUpcomingItem(u.string("id"), u.string("kind"), u.string("title"), u.string("detail"), u.long("dateMs"))
                },
            )
        }.getOrNull()
    }
}
